numpy/core/src/multiarray/item_selection.c

```diff
@@ -133,107 +243,50 @@ PyArray_TakeFrom(PyArrayObject *self0, PyObject *indices0, int axis,
-                            PyArray_Item_INCREF(tmp_src, PyArray_DESCR(self));
-                            PyArray_Item_XDECREF(dest, PyArray_DESCR(self));
-                            memmove(dest, tmp_src, itemsize);
-                            dest += itemsize;
-                            tmp_src += itemsize;
-                        }
-                    }
-                    else {
-                        memmove(dest, tmp_src, chunk);
-                        dest += chunk;
-                    }
-                }
-                src += chunk*max_item;
-            }
+        case 2:
+            res = npy_fasttake_impl(
+                    dest, src, indices_data, n, m, max_item, nelem, chunk,
+                    clipmode, 2, 0, dtype, axis);
             break;
-        case NPY_CLIP:
-            for (i = 0; i < n; i++) {
-                for (j = 0; j < m; j++) {
-                    tmp = ((npy_intp *)(PyArray_DATA(indices)))[j];
-                    if (tmp < 0) {
-                        tmp = 0;
-                    }
-                    else if (tmp >= max_item) {
-                        tmp = max_item - 1;
-                    }
-                    tmp_src = src + tmp * chunk;
-                    if (needs_refcounting) {
-                        for (k=0; k < nelem; k++) {
-                            PyArray_Item_INCREF(tmp_src, PyArray_DESCR(self));
-                            PyArray_Item_XDECREF(dest, PyArray_DESCR(self));
-                            memmove(dest, tmp_src, itemsize);
-                            dest += itemsize;
-                            tmp_src += itemsize;
-                        }
-                    }
-                    else {
-                        memmove(dest, tmp_src, chunk);
-                        dest += chunk;
-                    }
-                }
-                src += chunk*max_item;
-            }
+        case 4:
+            res = npy_fasttake_impl(
+                    dest, src, indices_data, n, m, max_item, nelem, chunk,
+                    clipmode, 4, 0, dtype, axis);
             break;
-        }
-        NPY_END_THREADS;
+        case 8:
+            res = npy_fasttake_impl(
+                    dest, src, indices_data, n, m, max_item, nelem, chunk,
+                    clipmode, 8, 0, dtype, axis);
+            break;
+        case 16:
+            res = npy_fasttake_impl(
+                    dest, src, indices_data, n, m, max_item, nelem, chunk,
+                    clipmode, 16, 0, dtype, axis);
+            break;
+        case 32:
+            res = npy_fasttake_impl(
+                    dest, src, indices_data, n, m, max_item, nelem, chunk,
+                    clipmode, 32, 0, dtype, axis);
+            break;
+        default:
+            res = npy_fasttake_impl(
+                    dest, src, indices_data, n, m, max_item, nelem, chunk,
+                    clipmode, itemsize, needs_refcounting, dtype, axis);
     }
```

Comment:
Can you move this entire function to `npy_fasttake`, which can have a very clear docstring of something like:
```
/* npy_fasttake_impl, instantiated repeatedly in different branches for optimization reasons */
```